### ISSUE TITLE
state transition performed twice when using optional state parameters

### ISSUE THREAD

**Issue by 3plusalpha** [angular-ui/ui-router#1396]
When optional state parameters are defined for a state, the state transition is performed twice as long as the optional parameters.

The $stateParams of a controller that was called twice look like this:
First: {offset: undefined, limit: undefined, sort: undefined, sortby: undefined}
Second: {offset: 0, limit: 20, sort: "asc", sortby: "title"} 

Note that all state parameters are optional and defined like this. 

```
.state({
  name: "list.products",
  parent: "list",
  url: "/products/:offset/:limit/:sortby/:sort/",
  params: {
    offset: { value: 0 },
    limit: { value: 20 },
    sort: { value: 'asc' },
    sortby: { value: 'title' }
  },
  templateUrl: "views/partials/product-search-result-item.html",
  controller: "ListProductsController"
})
```

When I use ui-sref without any parameter the controller is called twice.
<a ui-sref="list.products">Products</a>

Using all state parameters the controller is called only once.
<a ui-sref="list.products({offset:0,limit:20,sort:'asc',sortby:'title'})">Products</a>

IMHO this is a bug and I hope somebody can fix it as all controllers with optional parameters and API calls make API requests twice.

**Comment by nolandubeau** [comment]
Hi,  I'm seeing this when using version 0.2.13.  Instead of using ui-sref I have am using ng-click to call a function which then calls $state.go().  I can also confirm this when using $state.transitionTo(). 

ng-click="selectAlert(alert)

var selectAlert = function(alert){
$state.go('app.alerts.list.detail', {alertId: alert.id});
}

Does this bug still exist or am I using this incorrectly?

**Comment by realdah** [comment]
I had same problem like @nolandubeau . version is 0.2.11.
call $state.go in ng-click.  some $state parameters are null.


Seems to be fixed in 0.2.14.

**Comment by pranaydutta89** [comment]
i am still getting the same problem in 0.2.15 ,is the problem fixed or still persist ,please help me with this one

**Comment by bryanerayner** [comment]
I also am getting this problem in 0.2.15. I am using custom types in my solution.

The code which is resulting in the transition not being skipped, is in shouldSkipReload, line 3606.

   (locals === from.locals || ....)

Something is causing them to not be equal once the $locationChangeSuccess event is fired.

**Comment by christopherthielen** [comment]
Please make a plunker

**Comment by Gratouille** [comment]
+1, have the same problem in 0.2.15.

**Comment by pranaydutta89** [comment]
its working fine with 0.2.14,

**Comment by Gratouille** [comment]
I have a simple state with a resolve with a simple parameter. When I refresh the page (like url/state/paramId) the resolve is executed twice. 

0.2.14 0.2.10 0.2.15, same error, this is executed twice. It won't work neither with :paramId, {paramId} or ?paramId=id.

**Comment by blowsie** [comment]
I agree with pranaydutta89; for me using version 0.2.14 instead if 0.2.15 fixes my issue.
There does seem to be another issue here, which I am looking into.

**Comment by bryanerayner** [comment]
@christopherthielen I have a Plunkr which demonstrates what I'm trying to do to make some sort of dynamic parameters work in 0.2.15. http://plnkr.co/edit/Yeex1Rmd1nWQvdrxj87N?p=preview. It works to change parameters syncronously, but not via a $timeout.

**Comment by levivm** [comment]
Im getting this problem in 0.2.15. 

Using 0.2.14. fixed the problem to me.

**Comment by m2omou** [comment]
+1 Same for me, I had to go back to 0.2.14 to fix the problem

**Comment by petr0001** [comment]
+1 Same here

**Comment by m2omou** [comment]
Actually it's also happening in the version 0.2.14 even though this version looks more stable than the 0.2.15.

**Comment by zszep** [comment]
Can confirm this for version 0.2.15. When using nested views and a parameter is present in ui-sref, then the parent resolve function is fired for a second time. Removing the parameter makes it work as expected.

**Comment by rog1039** [comment]
+1 here too

**Comment by vroudge** [comment]
Can also confirm on 0.2.15

**Comment by christopherthielen** [comment]
@bryanerayner Thanks for the plunker, but this ticket is about "state transition performed twice when using optional state parameters".  Your plunker is demonstrating some other issue with reloadOnSearch (we only seem to support one level of reloadOnSearch at a time).  Dynamic parameters are scheduled for the 1.0 release.

Also, you should use inherit: true so you don't lose your current parameters.  
http://plnkr.co/edit/hMRjZSnLrjnCKnI5eqeT?p=preview

**Comment by christopherthielen** [comment]
To the rest of you who have +1'd or said "me too", I need you to reproduce this double-transition issue in a plunker.  This ticket is *closed*, the original issue was resolved, and I don't have any usable information to go on to fix the +1s

**Comment by yamateh** [comment]
Im getting this problem in 0.2.15.
